### ISSUE TITLE
Winnipeg cust # 075-4621; inv 075-16760- Code issue

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1673]
In GitLab by @kdjstudios on Apr 6, 2020, 09:42

**Submitted by:** "rich.soltoff@answernet.com" <rich.soltoff@answernet.com>

**Helpdesk:** http://servicedesk.answernet.com/profiles/ticket/2020-04-04-21421 

**Server:** Internal

**Client/Site:** Winnipeg

**Account:** 075-4621

**Issue:** 

The service code breakdown is using code 9998- Deposit Applied Debit Revenue, but the description on the invoice has nothing to do with a security deposit?

 

Please provide explanation.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 6, 2020, 09:43

I wrote:

> I took a quick look at the fees setup at the Winnipeg site and it would appear operations created a "TTY" additional fee that has the billing code 9998. Please let us know if you have any questions or concerns. 


@amishra @ranand 

Please look into this and explain how the site was able to create a billing code using one of the reserved SD code values.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 7, 2020, 05:40

@kdjstudios @tim.traylor 

Thank you for the update.

We looked into this and our findings are described below:

* I would like to inform you that the site has two additional fees with same code  **9998**

* One fee is Activated which name is **TTY**.
* The second fee is deactivated which name is **Deposit Applied Debit Revenue.**

* I had already informed you on my [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1730#note_56633) in ticket #1730

* Please let us know if we need to deactivate the fee too.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 7, 2020, 05:45

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 7, 2020, 10:01

@amishra @tim.traylor @Robleski 


As you can see in my comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1776#note_57653, we already know where the fee came from. However you failed to answered the question that was asked:

> Please look into this and explain how the site was able to create a billing code using one of the reserved SD code values.

EDIT:

Sorry I missed the following link on the word "commen"t and was looking at the link which went to the whole ticket number.  
* I had already informed you on my [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1730#note_56633) in ticket #1730

I will clarify that this fee was created before the SD fees were put into place. However. I did just test on Internal Production on the Demo site and I can create an additional fee without error using one of the SD fee numbers. Where is the validation and why is it not working?


![image](/uploads/e53d2b98b991f7631abaafac95ccda9d/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 8, 2020, 24:22

@kdjstudios @tim.traylor 

Thank you for the update.

* We would like to inform you that the validation functionality had developed only for the Activity code.
* After the long research, we found the ticket was #1479 . As you can see on my [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1479#note_45059) that we had given an estimate for the activity only.
* Although, we have reverted back the validation functionality because portland requested in ticket #1581
* Now, we are handling all the activity duplicate code on ticket #1628 

**IN SHORT: there is no validation functionality working for the activity or additional fee.**
 
Hope this help.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 15, 2020, 08:56

Marking for discussion, I believe we should be validating billing codes to insure we do not allow ops to produce duplicate codes.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 15, 2020, 15:12

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 12, 2020, 01:57

@kdjstudios @Robleski 

Please let us know what we need to do on this ticket?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 21, 2020, 08:59

plse create estimate, design, and AN will review and then give approval to start work.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 21, 2020, 09:02

AN to develop reqs and submit across.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 21, 2020, 13:26

We will want to update SA Billing so when a user attempts to save on the site edit page to validate the Billing Codes (Activities and Additional fees) that check if the billing code is already in use by another SD, Tax, holiday, activity or additional fee, etc.

Please let me know if anything else is needed or if there are any foreseeable concerns. Otherwise please provide and estimate.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 25, 2020, 04:06

@kdjstudios @Robleski 

Thank you for the update.

* As per my understanding, we handling the duplicate code issue on ticket #1628.
* Once we fix all duplicate code from each site(If any) then we can add the validation for all fields.
* Because if an invoice is created then we update the site holiday date(if any) and if we write the duplicate validation functionality then SAB unable to update the holiday date.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 26, 2020, 15:18

@amishra @Robleski 

I am confused, with how we are not able to add the validation first? May someone please explain in further detail what issue or error would occur. Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 28, 2020, 08:29

@kdjstudios @Robleski 

Currently, there are two ways in which the site gets updated.

1. From the edit site page: This scenario has been discussed
2. Automatic site update after application of the holiday fees.
   * In this scenario, the site's holiday date gets updated to next year after this year's holiday fees get applied to the invoices.
   * After duplicate code validation, the system will raise an error at the invoice creation if there will be any holiday fee on it.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 28, 2020, 10:01

@ranand @rranjan @amishra @Robleski 

So if I am understanding correctly we need to do the following in the order below. Please let me know if this makes sense or if there is a better way.

1. Provide a list of all sites and their duplicate fees for both internal and external. [This ticket]
2. Confirm and release updated Billing codes. [#1628 ticket] This will resolve internal server issues.
3. If applicable update any external duplicate fees. [This ticket]
4. Estimate and develop validation on the site edit page to prevent users from creating duplicate codes again. [This ticket]

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on May 29, 2020, 07:39

@kdjstudios  @Robleski 

Thanks for the update

> Provide a list of all sites and their duplicate fees for both internal and external.[This ticket]

Please find the attached list of all duplicate fees.

Internal Server

[Answernet_duplicate_code_on_site.csv](/uploads/82f241f73d5dc62278d24ec3571ace63/Answernet_duplicate_code_on_site.csv)

External Server

[Keener_duplicate_code_on_site.csv](/uploads/a6e5263719cc48ec1dc1566e6cd507a7/Keener_duplicate_code_on_site.csv)

>Estimate and develop validation on the site edit page to prevent users from creating duplicate codes again. \[This ticket\]


Do we need to give an estimate for the fix or we need to wait for ticket #1628 fixed?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 29, 2020, 09:09

@rranjan @ranand @amishra 

Please provide the estimate.

EDIT: I took a quick look at the lists, please also include if the codes are "active" on the list too. I have a strong feeling/hope most of these are deactivated already.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 1, 2020, 07:17

Awaiting Avaaz estimate.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 1, 2020, 07:20

@kdjstudios  @Robleski 

Thanks for the update.

Please find the updated list of CSV.

Internal server:

[1774_answernet_duplicate_code_on_site.csv](/uploads/757589d245fc4969d32a97fbeb472ab6/1774_answernet_duplicate_code_on_site.csv)


External server:

[1774_keener_duplicate_code_on_site.csv](/uploads/00a57c5880ad77fa818d4447fffb64eb/1774_keener_duplicate_code_on_site.csv)


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 1, 2020, 10:27

@Robleski @rranjan @ranand @amishra 

I think there is confusion on what we are trying to accomplish here, and I think that is due to me not being as precise with the request as needed. We need a list of **"Active" ONLY** codes that are duplicate on the site. Please group this list by site and sort by code number at each site. This means that there are two active codes that have the same billing code that operations is able to use currently. Please disregard any deactivated codes, as that is one of the purposes to deactivate a code is to be able to reuse that code for something else. I hope this makes sense, if not, please let us know of your concern or questions.

Also one thing I just thought of, is we will need to also prevent a user from "reactivating" a code if there is an already active code with the same billing code. This will be part of the step "4" from http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1776#note_59401.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Jun 2, 2020, 05:43

@kdjstudios @Robleski 

Please find the attached list of active duplicate codes

**Internal:**
[1774_answernet_active_duplicate_code_on_site.csv](/uploads/69d61eddafaadf7369a4bdd3377637a6/1774_answernet_active_duplicate_code_on_site.csv)

**External:**
[1774_keener_active_duplicate_code_on_site.csv](/uploads/fd544d58ec6424afc9b09590fde1fbf2/1774_keener_active_duplicate_code_on_site.csv)

We will provide the estimate tomorrow. Summary of requirements:
* Validation of unique code on site's active additional fees codes, activities codes, site security deposit codes, billing period codes, and late fee code.
* Any inactive fee can have a duplicate code.
* Cannot reactivate any inactive duplicate fee without removing duplication.

Please add any other requirement, if I missed it

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 2, 2020, 09:40

This was reviewed today and Gary will want us to reach out to each of the sites and inform them to deactivate one of the duplicate codes at the site level. I will start sending each of those emails out and post an update once completed.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 3, 2020, 01:02

@kdjstudios @Robleski 

Thank you for the update.

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|4|Worked with Concepts, implementation ok|6|High Impact|3|9|

##### Estimate Notes:
* Risk areas identified: `Report`
* Work plan:
   * Validation of unique code on site's active additional fees codes, activities codes, site security deposit codes, billing period codes, and late fee code.

* Test cases:
  * Testing with all billing code(add a new one code and same code)
  * Testing with activate/deactivate codes.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 3, 2020, 06:43

Avaaz is authorized to start work.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 9, 2020, 07:40

@kdjstudios @Robleski 

We have almost completed the development and most probably ready for your testing by tomorrow.

thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 10, 2020, 08:55

@amishra @rranjan @ranand @Robleski

Please let us know when this is ready for our testing.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 12, 2020, 07:23

@kdjstudios @Robleski 

We have setup branch on the dev server for your testing.

* We have validated on following code:
   * Activity(CDR FIELD)
   * Addition Fee
   * Security Deposit Fee
   * Billing Periods
   * Taxes

Please test it and let me know.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 15, 2020, 10:07

@amishra 

Looks good marking for Announcement.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 16, 2020, 01:47

@kdjstudios @Robleski 

Thank you for the testing and approval.

I would like to inform you that Please confirm with the client that all duplicate fee is deactivated or not?

Once resolve all duplicate fees we will release on the production server.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 16, 2020, 13:22

### Internal

* Atlanta - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-27497/conversation
* Billerica - - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-72713/conversation
* Billings - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-26461/conversation
* Columbus - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-25856/conversation
* El Paso - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-39456/conversation
* Hays - - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-45561/conversation
* Portland - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-18371/conversation
* Rockland - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-28749/conversation
* Santa Rosa - **AAVAZ** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-65913/conversation
* Stockton - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-82757/conversation
* Triad - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-91259/conversation
* NCSM - **DONE** - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-16-29747/conversation

### External

Towne - - https://www-servicedesk.answernet.com/profiles/ticket/2020-06-17-41654/conversation
Keener - All sites are terminated

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 16, 2020, 15:03

Santa Rosa wrote:

> the duplicate late charges won't allow me to deactivate either of them and the PayGo duplicates are not showing in the site settings.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 17, 2020, 06:58

@kdjstudios  @Robleski 

Thanks for the update.

We looked into the Santa Rosa site and find that following code are pre-defined:-

 *  Late Charge(999)
 *  Carry Forward Base Rate(9990)

so, it's not showing the button. Please confirm this so that we can disable them from the backend.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 17, 2020, 08:05

@rranjan 

Please ensure one of the following each listed on the spreadsheet below is deactivated

![image](/uploads/4e226b41c438f27be8f5fc48a9253c27/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 18, 2020, 01:32

@kdjstudios @Robleski 

Thank you for the update.

Please confirm all sites deactivated the duplicate code except the Santa Rosa site.

thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 18, 2020, 08:17

@amishra 

Is Santa Rosa Completed? Please let me know so I can confirm with Operations. Thanks.

EDIT: I will continue to update the post above as sites confirm they have updated. http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1776#note_60046

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 18, 2020, 08:26

@kdjstudios @Robleski 

No, I'll do it tomorrow on the production server.
**OR**
Do we need to do dev server first.

Please confirm .

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 18, 2020, 08:30

@amishra 

Please do this directly on Production since it should be a very quick and easy change.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 19, 2020, 24:55

@kdjstudios @Robleski 

Thank you for the Update.

* Deactivated one **999** code.
* Someone already changed **056 Miscellaneous Charge 9990** fee to **056 Miscellaneous Charge 9912** fee
* NCSM PAYG Receivable Transfer only one active fee we found, it might someone already changed.
* NCSM PAYG Receivable Transfer only one active fee we found, it might someone already changed.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 23, 2020, 07:22

Has this been applied in Prod?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 23, 2020, 12:08

@amishra @Robleski 

May we please run the check again to see which sites are left. I believe all but Towne, Billerica, and Hays we have heard back from ops stating they have completed the updates.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 24, 2020, 06:39

is this good to go? move to prod?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 24, 2020, 08:59

@Robleski 

I believe the code is ready. The only thing we are waiting on is ops/clients to manually fix the exiting duplicates.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 24, 2020, 13:20

@amishra @Robleski @ranand @rranjan 

As requested in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1776#note_60291 please rerun the check to confirm we are able to release this to production this weekend.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 25, 2020, 01:57

@kdjstudios @Robleski 

Thank you for the update.

We have run the script and found only three sites(Towne, Billerica, and Hays) are left as you mentioned above.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 25, 2020, 08:11

@amishra @Robleski 

Please provide the list from the script rerun so we may send to ops. Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 25, 2020, 08:39

@kdjstudios @Robleski 

Please find the below attached file.

#### INTERNAL
[internal.csv](/uploads/1905a0aee729e7db1d0cff6ddcc3e686/new_1774_answernet_duplicate_code_on_site_new_20202020.csv)

#### Hosted
[hosted.csv](/uploads/58b9663d402195583173620b80094018/new_1774_keener_duplicate_code_on_site_new_20202020.csv)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 26, 2020, 08:19

Both Billerica and Hays confirmed they have updated their sites now. We are waiting for Charlie at Towne to finish, I have been in communication with him today as he has concerns on how to deactivate fees.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 7, 2020, 13:27

Towne was able to confirm they were able to update the Billing Codes and deactivate the duplicates. They however found a few things that we need to check:

1. Please confirm that the current active activities match the ones on the HR and import. Please also make sure the activity name in the HR is being pulled from the activity at the site level.
2. The found that billing code 452 and 453 have the same calculation but they are supposed to be different. 453 should have be the sum of all metrics, it should not be subtracting anything. 


Please make sure the above are completed. Please also confirm there are no sites then with any duplicates so we can release this on saturday.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jul 8, 2020, 02:08

@kdjstudios @Robleski 

> Please confirm that the current active activities match the ones on the HR and import. Please also make sure the activity name in the HR is being pulled from the activity at the site level.

* After release #524 we have created one model(Switch Activity) which is associated with the switch.  
* We have defined all activities according to switch.

* If the switch has switch activities then we show the HR report activities from the switch activities else show the site activities.


> The found that billing code 452 and 453 have the same calculation but they are supposed to be different. 453 should have be the sum of all metrics, it should not be subtracting anything.

* Yes , Kyle's you are right. We looked into the posting script they are not subtracting for code `453`.

**Do we need to create new switch version to fixed the UI calculation issue for code `453`?**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 8, 2020, 10:31

@amishra 

Yes please make the new switch version with the updated calculation. Also as requested confirm that the active activities on the site are the ones linked to the switch/HR. If they do not match update those as well.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jul 13, 2020, 06:20

@kdjstudios  @Robleski 

Thanks for the update.

We have completed its development and setup branch on the merge server for your testing. 

Also,

While development we have noticed that "411" activity code is repeated.So should we remove the duplicate code or what we need to do in this scenario? 

![image](/uploads/f8e0d11627f36f2bb5743a0aeb795513/image.png)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 13, 2020, 08:15

@rranjan @ranand @amishra 

Did the client not deactivate one of the 411 odes when the deactivated all other duplicates? It looks to me like they had removed all duplicates. Now if your team created duplicates on the HR report, then please fix that, since operations/client does not have access to that.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jul 14, 2020, 04:14

@kdjstudios  @Robleski 

Thanks for the update.

We have fixed the duplicate code and setup branch on the merge server for your testing.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 14, 2020, 11:40

Marking for ready. Link to CERB: https://www-servicedesk.answernet.com/profiles/ticket/2020-06-17-41654/conversation

EDIT: Please release this update overnight. I do not want to wait until the weekend and have these effect any of the clients billing any longer.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jul 15, 2020, 01:34

@kdjstudios @Robleski 

Thank you for the update.

* We have released the posting script & switch part on the production server.
* We still need to release the second part(duplicate code) .

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 15, 2020, 08:33

@amishra @Robleski 

Why was the full update not released as requested?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jul 16, 2020, 02:13

@kdjstudios @Robleski 

We have successfully released on both of the production servers

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 20, 2020, 11:12

closed
